packages/core/src/mixer.ts

```diff
@@ -25,42 +23,129 @@ const comparator = (a: HeapElement, b: HeapElement): number => {
  */
 export class Mixer<Chain extends HoprCoreConnector> {
   private queue: Heap<HeapElement>
+  private timeout?: NodeJS.Timeout
+  private defer: DeferredPromise<void>
+  private deferEnd: DeferredPromise<void>
+  private endPromise: Promise<void>
+  private _done: boolean
 
   public WAIT_TIME = MAX_PACKET_DELAY
 
   constructor(private incrementer = Date.now) {
     this.queue = new Heap(comparator)
-  }
 
-  push(p: Packet<Chain>) {
-    this.queue.push([this.getPriority(), p])
+    this.defer = Defer<void>()
+    this.deferEnd = Defer<void>()
+
+    this._done = false
+
+    this.endPromise = this.deferEnd.promise.then(() => {
+      this._done = true
+    })
   }
 
-  // Can we pop an element?.
-  poppable(): boolean {
-    log(`Mixer has ${this.queue.length} elements`)
-    if (!this.queue.length) {
-      return false
+  /**
+   * Adds a packet to the mixer
+   * @param p
+   */
+  public push(p: Packet<Chain>) {
+    if (this._done) {
+      throw Error(`Mixer has ended. Could not accept any further messages.`)
     }
-    return this.queue.peek()[0] < this.due()
-  }
+    const newPriority = this.getPriority()
+    const topPriority = this.queue.length > 0 ? this.queue.peek()[0] : Number.MAX_SAFE_INTEGER
 
-  pop(): Packet<Chain> {
-    if (!this.poppable()) {
-      throw new Error('No packet is ready to be popped from mixer')
+    if (newPriority < topPriority) {
+      this.resetTimeout(newPriority)
     }
-    return this.queue.pop()[1]
+
+    log(`Added 1 packet to the mixer`)
+    this.queue.push([newPriority, p])
+  }
+
+  /**
+   * Get the number of packets that are currently
+   * in the mixer
+   */
+  public get length() {
+    return this.queue.length
+  }
+
+  /**
+   * Adjust the internal timeout with latest priority
+   * @param newPriority
+   */
+  private resetTimeout(newPriority: number) {
+    clearTimeout(this.timeout)
+    this.timeout = setTimeout(() => {
+      this.defer.resolve()
+    }, newPriority - Date.now())
   }
 
   notEmpty(): boolean {
     return this.queue.length > 0
   }
 
-  private due(): number {
-    return this.incrementer()
+  /**
+   * Ends the mixer.
+   */
+  public end(): void {
```

Comment:
Why does a mixer need an end?

Reply:
To be able to stop mixing without producing a hanging Promise

Reply:
Why? we can just not call pop() after it needs to stop.

Reply:
pop() returns a Promise that resolves once the next packet is ready. If there is no packet because we stopped mixing, then we will wait forever

Reply:
And end makes pop() resolve with an empty value? This means pop() needs to be checked afterwards.

It's better to make the mixer be unaware of 'ending'.

Reply:
> It's better to make the mixer be unaware of 'ending'.

Yes, this indeed a little bit hacky. Not sure how to implemenet this because once calling, we don't know whether we will get a packet or the mixer has ended meanwhile.

We could throw an exception. But this exception need to be caught - which is similar as handling an "end" message

Reply:
we should get rid of 'ending' at all in the mixer. It's one more state that is completely unneccessary.

Reply:
If the solution were to have hanging promises, then I disagree here. Knowingly producing hanging promises is one of the worst code styles.